### PR TITLE
feat: add stats for bad events in fb_custom_audience

### PR DESCRIPTION
**Pull request by ujjwal-ab**
## Description of the change

Add stats for bad events in fb_custom_audience

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

**Comment by Gauravudia** [comment]
@ujjwal-ab You need to add the stats in the `prometheus.js` file as well.

**Review by koladilip — DISMISSED** [review]
destinationId destination_id destination we are using different labels, something we need to standardise

**Review by sanpj2292 — APPROVED** [review]

File: src/v0/destinations/fb_custom_audience/util.js [hunk 1]
```diff
@@ -126,43 +127,77 @@ const ensureApplicableFormat = (userProperty, userInformation) => {
...
+      );
+
+      if (dataElement[dataElement.length - 1]) {
+        nullUserData = false;
+        nullEvent = false;
       }
     });
+
+    if (nullUserData) {
+      stats.increment('fb_custom_audience_event_having_all_null_field_values_for_a_user', {
```
Comment (sanpj2292) [inline src/v0/destinations/fb_custom_audience/util.js:186]: [Q1] Will the stat add value if we know which field is null ?
[Q2] Are the user-fields going to a lot ? I was asking in terms of cardinality

Comment (sanpj2292) [reply src/v0/destinations/fb_custom_audience/util.js:186]: >Are the user-fields going to a lot ?

The intention is to know if these user-fields are fixed but the numbers are high or if the user-fields are very random. With that we can estimate the cardinality

**Review by sanpj2292 — APPROVED** [review]

File: src/v0/destinations/fb_custom_audience/util.js (hunk 1 above)
Comment (ujjwal-ab) [reply src/v0/destinations/fb_custom_audience/util.js:186]: 1. We just want to get the count of the event in which all the fields are null, doesn't matter which field is null here.
Didn't get the second question.

Comment (sanpj2292) [reply src/v0/destinations/fb_custom_audience/util.js:186]: >We just want to get the count of the event in which all the fields are null, doesn't matter which field is null here.

I understand, but if I have a label which says that this field that is sent is `null`, would it add value ? -- This is my question

Second question was, how many user-fields would be sent ? Can it be randomly sent by user or is it a fixed set of values that should be set as part of the payload ?!

**Review by koladilip — DISMISSED** [review]

File: src/v0/destinations/fb_custom_audience/util.js [hunk 2]
```diff
@@ -126,43 +127,78 @@ const ensureApplicableFormat = (userProperty, userInformation) => {
...
 
+const getUpdatedDataElement = (dataElement, isHashRequired, eachProperty, updatedProperty) => {
+  let tmpUpdatedProperty = updatedProperty;
+  if (isHashRequired && eachProperty !== 'MADID' && eachProperty !== 'EXTERN_ID') {
+    // for MOBILE_ADVERTISER_ID, MADID,EXTERN_ID hashing is not required ref: https://developers.facebook.com/docs/marketing-api/audiences/guides/custom-audiences#hash
+    if (tmpUpdatedProperty) {
+      tmpUpdatedProperty = `${tmpUpdatedProperty}`;
+      dataElement.push(sha256(tmpUpdatedProperty));
+    } else {
+      dataElement.push(null);
```
[the remark this answers is not in the document]
Comment (ujjwal-ab) [reply src/v0/destinations/fb_custom_audience/util.js:138]: Yes, if the user is not providing the values for the properties that they have selected to send. That's why we are sending null to those fields.

File: src/v0/destinations/fb_custom_audience/util.js [hunk 3]
```diff
@@ -126,43 +127,78 @@ const ensureApplicableFormat = (userProperty, userInformation) => {
   return updatedProperty;
 };
 
+const getUpdatedDataElement = (dataElement, isHashRequired, eachProperty, updatedProperty) => {
+  let tmpUpdatedProperty = updatedProperty;
+  if (isHashRequired && eachProperty !== 'MADID' && eachProperty !== 'EXTERN_ID') {
```
[the remark this answers is not in the document]
Comment (ujjwal-ab) [reply src/v0/destinations/fb_custom_audience/util.js:132]: Which one you are referring to simplify, we have only two conditions. Have refactored the code to remove branchings as I was getting code smell warning. If I will do some more branching will get code smell for cognitive complexity.

[the remark this answers is not in the document]
Comment (koladilip) [reply src/v0/destinations/fb_custom_audience/util.js:132]: can you add comment on the condition block, explaining what is the condition means

File: src/v0/destinations/fb_custom_audience/util.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (koladilip) [reply src/v0/destinations/fb_custom_audience/util.js:138]: sending nulls is not causing any issues right? incase pinterest we got some issues right

[the remark this answers is not in the document]
Comment (ujjwal-ab) [reply src/v0/destinations/fb_custom_audience/util.js:139]: No, it is not causing issue. fb accepts null values

File: src/v0/destinations/fb_custom_audience/util.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (ujjwal-ab) [reply src/v0/destinations/fb_custom_audience/util.js:133]: Added comments

File: src/v0/destinations/fb_custom_audience/util.js [hunk 4]
```diff
@@ -129,6 +129,7 @@ const ensureApplicableFormat = (userProperty, userInformation) => {
 
 const getUpdatedDataElement = (dataElement, isHashRequired, eachProperty, updatedProperty) => {
   let tmpUpdatedProperty = updatedProperty;
+  // if hashRequired is true and the property name is not (MADID or EXTERN_ID) hash the value if it is defined else send null
```
Comment (koladilip) [inline src/v0/destinations/fb_custom_audience/util.js:132]: this comment is not helpful as it is direct translation of code, you can remove it, it is fine.

**Review by koladilip — DISMISSED** [review]

**Review by koladilip — APPROVED** [review]

**Review by sanpj2292 — APPROVED** [review]

**Pull request merged**